### PR TITLE
typescript phenyl-central-state

### PR DESCRIPTION
**Pull request by nicholaslee119**
typescript phenyl-central-state

### CODE REVIEW

**Comment by nicholaslee119** [comment]
the type of DBClient should be improved by changing `Entity`  to `EntityWithMeteInfo`

File: modules/central-state/package.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  "repository": "phenyl-js/phenyl",
+  "license": "Apache License 2.0",
+  "author": "Shin Suzuki <shinout310@gmail.com>, Nicholas Lee <nicholaslee119@gmail.com>",
+  "files": [
+    "lib",
+    "LICENSE",
+    "README.md"
+  ],
+  "main": "lib/index.js",
+  "types": "lib/index.d.js",
```
Comment (Naturalclar) [inline modules/central-state/package.json]: types should be changed to `lib/index.d.ts`

Comment (nicholaslee119) [reply modules/central-state/package.json]: Fixed!

File: modules/central-state/package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  "devDependencies": {
+    "babel-register": "^6.24.1",
+    "mocha": "^5.0.5",
+    "power-assert": "^1.4.4"
+  },
+  "dependencies": {
+    "@sp2/format": "~1.2.3",
+    "@sp2/updater": "~1.2.3",
+    "oad-utils": "~0.3.0",
```
Comment (Naturalclar) [inline modules/central-state/package.json]: remove oad-utils, it is contained in sp2

File: modules/central-state/package.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    "@phenyl/interfaces": "*",
```
Comment (Naturalclar) [inline modules/central-state/package.json]: change version number to `1.0.0`

File: modules/central-state/src/entity-client.ts [hunk 4]
```diff
@@ -35,19 +35,22 @@ import type {
...
+export type PhenylEntityClientOptions<M extends GeneralReqResEntityMap> = {
   validatePushCommand?: PushValidation<M>,
 }
 
 /**
  * Validate PushCommand only when masterOperations are found.
  * masterOperations are not found when the versionId in PushCommand is over 100 commits older, as entity saves only 100 commits.
  */
-function validWhenDiffsFound(command: PushCommand<*>, entity: Entity, masterOperations: ?Array<UpdateOperation>) {
+function validWhenDiffsFound(command: PushCommand<any>, entity: Entity, masterOperations: Array<GeneralUpdateOperation> | null | undefined) {
```
Comment (Naturalclar) [inline modules/central-state/src/entity-client.ts]: add TODO comment to define `any`

File: modules/central-state/src/entity-client.ts [hunk 5]
```diff
@@ -198,8 +202,8 @@ export class PhenylEntityClient<M: EntityMap> implements EntityClient<M> {
   /**
    *
    */
-  async delete<N: $Keys<M>>(command: DeleteCommand<N>): Promise<DeleteCommandResult> {
-    return { ok: 1, n: await this.dbClient.delete(command) }
+  async delete<N extends Key<M>>(command: DeleteCommand<N>): Promise<DeleteCommandResult> {
```
Comment (Naturalclar) [inline modules/central-state/src/entity-client.ts]: change N to EN

File: modules/central-state/src/session-client.ts [hunk 6]
```diff
@@ -1,15 +1,16 @@
...
   Session,
   SessionClient,
-} from 'phenyl-interfaces'
+  PreEntity,
+} from '@phenyl/interfaces'
+
+type Id = string
 
 type PhenylSessionEntityMap = {
   _PhenylSession: Session
```
Comment (Naturalclar) [inline modules/central-state/src/session-client.ts]: TODO: Session can receive type parameter.

File: modules/central-state/src/session-client.ts [hunk 7]
```diff
@@ -21,14 +22,14 @@ type PhenylSessionEntityMap = {
 export class PhenylSessionClient implements SessionClient {
   dbClient: DbClient<PhenylSessionEntityMap>
 
-  constructor(dbClient: DbClient<*>) {
+  constructor(dbClient: DbClient<any>) {
```
Comment (Naturalclar) [inline modules/central-state/src/session-client.ts]: change any to PhenylSessionEntityMap

File: modules/central-state/src/versioning.ts [hunk 8]
```diff
@@ -1,26 +1,27 @@
-// @flow
 import {
   normalizeUpdateOperation,
   mergeUpdateOperations,
 } from 'oad-utils/jsnext'
```
Comment (Naturalclar) [inline modules/central-state/src/versioning.ts]: use sp2

File: modules/central-state/src/versioning.ts [hunk 9]
```diff
@@ -1,26 +1,27 @@
...
   MultiValuesCommandResult,
   PullQueryResult,
   PushCommandResult,
   QueryResult,
   SingleQueryResult,
-  UpdateOperation,
-} from 'phenyl-interfaces'
+} from '@phenyl/interfaces'
+
+import { UpdateOperation, UpdateOperator } from '@sp2/format'
```
Comment (Naturalclar) [inline modules/central-state/src/versioning.ts]: use sp2

File: modules/central-state/package.json [hunk 10]
```diff
@@ -17,15 +17,11 @@
...
   },
   "dependencies": {
-    "@sp2/format": "~1.2.3",
-    "@sp2/updater": "~1.2.3",
-    "oad-utils": "~0.3.0",
-    "@phenyl/interfaces": "*",
-    "@phenyl/utils": "*"
+    "@phenyl/interfaces": "^1.0.0",
+    "@phenyl/utils": "^1.0.0",
+    "sp2": "^1.2.0"
```
Comment (Naturalclar) [inline modules/central-state/package.json]: 1.3.0?

**Pull request merged**